cpp/src/strings/utilities.cuh

```diff
@@ -54,23 +56,56 @@ __device__ inline char* copy_string( char* buffer, const string_view& d_string )
  * @stream Stream to use for any kernel calls.
  * @return offsets child column for strings column
  */
-template <typename Iterator>
-std::unique_ptr<column> make_offsets( Iterator begin, Iterator end,
+template <typename InputIterator>
+std::unique_ptr<column> make_offsets_child_column( InputIterator begin, InputIterator end,
     rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
     cudaStream_t stream = 0)
 {
-    CUDF_EXPECTS(begin <= end, "Invalid iterator range");
+    CUDF_EXPECTS(begin < end, "Invalid iterator range");
     auto count = thrust::distance(begin, end);
     auto offsets_column = make_numeric_column(
           data_type{INT32}, count + 1, mask_state::UNALLOCATED, stream, mr);
     auto offsets_view = offsets_column->mutable_view();
     auto d_offsets = offsets_view.template data<int32_t>();
+    // Using inclusive-scan to compute last entry which is the total size.
+    // Exclusive-scan is possible but will not compute that last entry.
+    // Rather than manually computing the final offset using values in device memory,
+    // we use inclusive-scan on a shifted output (d_offsets+1) and then set the first
+    // offset values to zero manually.
     thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), begin, end,
                            d_offsets+1);
     CUDA_TRY(cudaMemsetAsync(d_offsets, 0, sizeof(int32_t), stream));
     return offsets_column;
 }
 
+/**
+ * @brief Utility to create a null mask for a strings column using a custom function.
```

Comment:
I think this will be superseded by a `valid_if` API that returns a `device_buffer` yes?

Can you leave a `TODO` note to replace it once `valid_if` is updated?

Reply:
Yep. The TODO is on https://github.com/rapidsai/cudf/blob/576d0b6a84ffbc38d5bf4a230c150c33c47004a6/cpp/src/strings/utilities.cuh#L105
I just refactored some code so the TODO should now be in one place in this file.